### ISSUE TITLE
Interact with canvas pixel data using Emscripten 

### ISSUE THREAD

**Issue by fishinTyme** [emscripten-core/emscripten#4140]
Is it possible to pass an array buffer of image data populated in "regular javascript" with the getImageData command, and pass that data off to Emscripten compiled javascript?

I'm basically wondering if you can leverage Emscripten's ability to write asm.js code to improve canvas pixel manipulation performance. Thank you!

**Comment by kripken** [comment]
Yes, however, you would need to copy the data into emscripten's memory (which is a single ArrayBuffer), and copy it out. So the question is whether the fast execution of asm.js is worth the extra copies.

In my experience it really depends: if you're doing a small simple loop on the pixels (like making everything brighter) then normal JS is already very fast, and the copies are not worth it. However, if you do a more complex operation, and especially if it's a large codebase, then asm.js quickly becomes worth the copies, since it can be many times faster.

**Comment by fishinTyme** [comment]
Wow thank you for your quick response! I wasn't expecting anyone to reply in less than 5 minutes. 

I am fairly confident that my program will benefit from the speeds of asm.js despite the memory copies. However I have another question based on what you wrote. My project does not just work with 1 image, it works with several. 

In my project I have many images of the same size. The javascript program compares them with various techniques pixel to pixel, and outputs several images at 60 fps (request animation frame) displaying the differences of the pictures using several comparison formulas. The final output of each "frame" is the same size as the images going in. 

So a simple example of what I would have using 2 images could like this (in standard english). 

image 1 is 500px by 500px, 
image 2 is also 500px by 500px, 
and the program generates "image3" showing the differences between the 2 input images. The dimensions of image 3 is also 500px by 500px.
Then 16 ms later it produces "image4" showing a new view on the differences. Image 4 is also 500px by 500px and so on.

So in my mind this is what I would like to setup using emscripten. 

Pass all the input image data into emscriptens memory. Have the emscripten function run at 60fps (if I have to use setTimeout or the C++ equivalent when programming this logic in C++ that's okay) then have the function spit out just 1 frame of rendered data every iteration of the function. 

In other words if the program has 5 images going into it, then the Heap memory is going to be 6 times larger than a single image (I say 6 because 1 of the times is dedicated space to hold the rendered data). However I only want to recover from emscripten that small portion of the memory containing the rendered data. 

I hope all that makes sense. Sorry for the long post, but I'm really interested in using emscripten to leverage the speeds of asm.js. I just can't find enough documentation online to help me understand the best way to set something like this up. Thank you for you help!

**Comment by kripken** [comment]
Yes, that all sounds reasonable. You can do one set of copies initially, then for each frame, read out just the output frame. Should be quite fast.

In overview, you'd probably want to do something like this:

1. Write some JS that mallocs enough room for the buffers, then copies into them. malloc returns an int, which is the location in memory. Just copy into HEAP8 which is a typed array viewing the heap as 8-bit values, at the right location.
2. Call an exported C method, giving it the pointer(s) to those buffers. Then that C code can do all the processing you need.
3. Use `emscripten_set_main_loop` to run a loop, and in each frame it will call a C method you told it. It can then call out to JS, notifying JS that it can read a frame of data (again, reading from HEAP8). The call to JS can be done using an `EM_ASM` call for example.

Examples of all of those things should be here: http://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html

There is more than one way to do things, though. For example you can just do a `setTimeout` yourself for the frame callbacks if you prefer that, and call into C from there, etc.

**Comment by bodokaiser** [comment]
@fishinTyme Did you got this running? I also would like to pass an image `ArrayBuffer` down to emscripten c code.

**Comment by rishantagarwal** [comment]
@fishinTyme @bodokaiser I was trying to do something similar. Any updates on this if it worked ?
